### PR TITLE
be/jvm: First version of a JVM bytecode backend.

### PR DESCRIPTION
**Pull request by fridis**
There are two new options to the `fz` command:

 - `-jvm` creates class files and immediately loads them via a ClassLoader and
   executes the code, similar to `-interpreter`, and

  - `-classes` that creates a directory `fuzion_generated_classes` that
    contains class files.

Biggest restrictions right now

 - Only part of the intrinsics are implemented

 - value types are not copied when assigned such that they have reference
   semantics.

 - no threads

 - no abort for effects

 - many other minor things, many marked with NYI in the sources

### CODE REVIEW

File: src/dev/flang/be/jvm/classfile/ClassFile.java~
Comment (michaellilltokiwa) [inline src/dev/flang/be/jvm/classfile/ClassFile.java~:1]: this file should probably be removed

File: src/dev/flang/be/jvm/classfile/ClassFileConstants.java [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+{
+
+
+  /*----------------------------  constants  ----------------------------*/
+
+
+  public static final byte[] MAGIC = new byte[] { (byte) 0xCA, (byte) 0xFE, (byte) 0xBA, (byte) 0xBE };
+
+  public static byte[] VERSION_JDK_1_0 = new byte[] { 0, 0, 0, 45 };
+  public static byte[] VERSION_JDK_1_1 = new byte[] { 0, 3, 0, 45 };
```
Comment (michaellilltokiwa) [inline src/dev/flang/be/jvm/classfile/ClassFileConstants.java:52]: the 3 in `0, 3, 0, 35` seems odd?

Comment (fridis) [reply src/dev/flang/be/jvm/classfile/ClassFileConstants.java:52]: This is correct, in the early days of Java the minor version was used.

File: src/dev/flang/be/jvm/runtime/Intrinsics.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+  public static boolean f64_infix_gt                (double a, double b) { return                 (a >   b); }
+  public static long    f64_as_i64_lax              (double a          ) { return (long)          (      a); }
+  public static float   f64_as_f32                  (double a          ) { return (float)         (      a); }
+  public static long    f64_cast_to_u64             (double a          ) { return Double.doubleToLongBits(a); }
+  public static boolean f32_type_is_NaN             (float  a          ) { return Float.isNaN       (         a); }
+  public static boolean f64_type_is_NaN             (double a          ) { return Double.isNaN      (         a); }
+  public static float   f32_type_acos               (float  a          ) { return (float) Math.acos (         a); }
+  public static float   f32_type_asin               (float  a          ) { return (float) Math.asin (         a); }
+  public static float   f32_type_atan               (float  a          ) { return (float) Math.atan (         a); }
+//public static float   f32_type_atan2              (float  a, float  b) { return (float) Math.atan2(a,       b); }
```
Comment (michaellilltokiwa) [inline src/dev/flang/be/jvm/runtime/Intrinsics.java:219]: Is there a reason atan2 is still commented out?

File: src/dev/flang/be/jvm/classfile/Expr.java [hunk 3]
```diff
@@ -0,0 +1,1062 @@
...
+      {
+        public String toString() { return "checkcast"; }
+        public JavaType type()
+        {
+          return type;
+        }
+        public byte[] byteCode(ClassFile cf)
+        {
+          return bc(O_checkcast, cf.cpClass(type.descriptor2()));
+          // return bc(O_checkcast, cf.cpClass(type.className()));
```
[the remark this answers is not in the document]
Comment (fridis) [reply src/dev/flang/be/jvm/classfile/Expr.java:912]: was forgotten.

File: src/dev/flang/be/jvm/classfile/Expr.java [hunk 4]
```diff
@@ -0,0 +1,1062 @@
...
+      {
+        public String toString() { return "endless_loop"; }
+        public JavaType type()
+        {
+          return ClassFileConstants.PrimitiveType.type_void;
+        }
+        public byte[] byteCode(ClassFile cf)
+        {
+          return bcsigned(O_goto, 0);
+          // return bc(O_checkcast, cf.cpClass(type.className()));
```
[the remark this answers is not in the document]
Comment (fridis) [reply src/dev/flang/be/jvm/classfile/Expr.java:935]: was forgotten.

File: src/dev/flang/be/jvm/runtime/Intrinsics.java [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+  public static int     f32_type_min_exp            (                  ) { return Float.MIN_EXPONENT;             }
+  public static float   f32_type_sin                (float  a          ) { return (float) Math.sin  (         a); }
+  public static float   f32_type_sinh               (float  a          ) { return (float) Math.sinh (         a); }
+  public static float   f32_type_square_root        (float  a          ) { return (float) Math.sqrt ((double) a); }
+  public static float   f32_type_tan                (float  a          ) { return (float) Math.tan  (         a); }
+  public static float   f32_type_tanh               (float  a          ) { return (float) Math.tanh (         a); }
+  public static double  f64_type_acos               (double a          ) { return         Math.acos (         a); }
+  public static double  f64_type_asin               (double a          ) { return         Math.asin (         a); }
+  public static double  f64_type_atan               (double a          ) { return         Math.atan (         a); }
+//public static double  f64_type_atan2              (double a, double b) { return         Math.atan2(a,       b); }
```
[the remark this answers is not in the document]
Comment (fridis) [reply src/dev/flang/be/jvm/runtime/Intrinsics.java:237]: I think was confused why this is implemented in Fuzion and not an intrinsic, will remove it.

**Pull request merged**